src/debug.cc

```diff
@@ -880,6 +965,54 @@ ForceAlert(std::ostream& s)
     return s;
 }
 
+/* DebugMessage */
+
+DebugMessage::DebugMessage(const int aSection, const int aLevel, const char *format, va_list args):
+    section(aSection),
+    level(aLevel)
+{
+    // The two paranoid(?) termination lines below are meant for vsnprintf()
+    // implementations that do not terminate on various kinds of errors.
+    // TODO: Unify all Squid vsnprintf() calls, probably using std::vsnprintf().
+    image[0] = '\0';
+    (void)vsnprintf(image, sizeof(image), format, args);
+    image[sizeof(image)-1] = '\0';
```

Comment:
I think this is not just 'paranoid' but also wrong. If we do not trust that vsnprintf() terminates the written buffer, we should get the number of effectively written characters from it and nullify the next byte, instead of nullifying the last byte of this huge buffer.

Reply:
AFAICT, the assumption here is that these two vsnprintf() implementation problems are possible:

1. vsnprintf() appends nothing (e.g. due to some format error), leaving the buffer "as is"
2. vsnprintf() appends sizeof(buf) characters, leaving the buffer full but unterminated

I believe the second problem has been observed in practice. I am not sure about the first one, but it is certainly possible to imagine such a rarely-triggered bug. The two "paranoid" code lines address these two problems, respectively.

> we should get the number of effectively written characters and nullify the next byte

The approach you describe is necessary only if there is a different problem:

3. vsnprintf() appended less than sizeof(buf) characters, left the buffer unterminated, and returned the number of characters appended (as it does on all successful operations).

AFAICT, the assumption here is that no vsnprintf() implementation is that broken. I think that is a reasonable assumption!


----

... where `sizeof(buf)` is a mnemonic for the second vsnprintf() parameter.